Missing recordings blog post

rcmarron opened the pull request:
## Changes

Blog post on how we solved the missing recordings issue.

## Checklist
- [x] Titles are in [sentence case](https://apastyle.apa.org/style-grammar-guidelines/capitalization/sentence-case)
- [x] Feature names are in [title case](https://apastyle.apa.org/style-grammar-guidelines/capitalization/title-case)
- [x] Words are spelled using American English
- [x] I have checked out our [style guide](https://github.com/PostHog/posthog.com/blob/master/STYLEGUIDE.md)

joethreepwood commented on contents/blog/missing-recordings.md:
```diff
@@ -0,0 +1,45 @@
+# The mystery of missing recordings at PostHog
+
+Session recordings is one of PostHog's core features. It allows customers to see their website being used, mouse clicks and all. It's really cool, but, until recently, we had one big issue with the feature. Customers would often tell us recordings were missing, and we didn't know why.
```
```suggestion
Session Recordings is one of PostHog's core features. It allows customers to see their website being used, mouse clicks and all. It's really cool, but, until recently, we had one big issue with the feature. Customers would often tell us recordings were missing, and we didn't know why.
```

andyvan-ph commented on contents/blog/missing-recordings.md:
```diff
@@ -0,0 +1,56 @@
+categories: ["Engineering"]
+---
+
+Session Recording is one of PostHog's core features. It allows customers to see their website being used, mouse clicks and all. It's really cool, but, until recently, we had one big issue with the feature. Customers would often tell us recordings were missing, and we didn't know why.
```
I think the intro should talk a little more about the Session Recording feature. How does it work, how do people use it etc. We want to talk up the feature as much as possible.

andyvan-ph commented on contents/blog/missing-recordings.md:
```diff
@@ -0,0 +1,56 @@
+
+## Slow moving metrics
+
+Because this fix went into posthog-js, and different customers upgrade at different speeds, it took a while to see the full effect of the fix in our metrics. However, after about a month of waiting, the 15% missing recordings metric dropped to <1%.
```
I feel like we're missing a conclusion here:

1) What did we learn from fixing this problem?
2) What feedback, if any, did we receieve from customers?

rcmarron replied:
100% agree. Just added a conclusion that touched on the other quality improvements that we made, and the impact we've seen on usage.

andyvan-ph commented:
Thanks @rcmarron. This is a really good starting point. A couple of comments from me on the intro and conclusion. Let me know what you think.

andyvan-ph commented on contents/blog/missing-recordings.md:
```diff
@@ -0,0 +1,58 @@
+categories: ["Engineering"]
+---
+
+[Session Recording](product/session-recording) is one of PostHog's core features. It allows customers to see their website being used, mouse clicks and all. It's really cool, but, until recently, we had one big issue with the feature. Customers would often tell us recordings were missing, and we didn't know why.
```
Am I correct in thinking the feature was still in beta when this problem was highlighted? We should say that if it's the case.

rcmarron replied:
Good call - snuck in a line about moving it out of beta.

andyvan-ph commented on contents/blog/missing-recordings.md:
```diff
@@ -0,0 +1,58 @@
+
+To figure out what was causing this error, we looked at data that had been uploaded that was just below our size threshold of 20mb. It looked like large data URLs of images were the cause.
+
+To test this out, we made a filter that would remove data URLs from any payload that exceeded a 5mb threshold, and it would replace the image with a simple SVG.
```
Just looking at the PR, I think it would nice to add a little context over the SVG decision. There's a nice bit of back and forth and we can use the embedded images to add some visuals to the piece.

rcmarron replied:
Great idea. Pulled some of that discussion out into the post.

andyvan-ph commented:
Throwing this out to the @PostHog/blog-reviewers team for more feedback. Some specific things to think about:

- Will existing users find this blog interesting / useful?
- Are there any technical elements we could add to appeal to engineers?
- Any suggestions RE: the headline would be welcome

**IMPORTANT:** Please don't merge when you've finished reviewing.

rcmarron commented:
Personally, I'm a bit lukewarm on this blog post right now.

I wonder if there's a better post to be written that's more of a "Check out all these cool things we added to recordings". The contents of this blog post would be summarized into one paragraph of it, but the post would also highlight other big improvements that we made to the feature.

Curious about your thoughts on this @andyvan-ph? I don't think it'd be too much to pull it together

pauldambra commented:
@rcmarron did we use PostHog to measure the metrics? I wonder if the hook is going meta. Here's how we used PostHog to measure usage of PostHog and make Session Recordings awesome...

neilkakkar commented:
Love this!

> I wonder if there's a better post to be written that's more of a "Check out all these cool things we added to recordings". 

Based on what I'd read, this current post is 10x better than the above^

Love the technical details and links to PRs/issues, A+!

> Are there any technical elements we could add to appeal to engineers?

I'd _love_ to see a quick summary of how session recording actually works with rrweb, what a full snapshot means, etc. etc.

Right now, a possible problem is that this post makes sense only to people who: (1) have a basic understanding of how session recording works; & (2) Are interested in hearing war stories.

We could easily change this intersection into a union by providing (1) in the post itself.

I'd also love hearing about predictions why the last 1% still exists: Can it all be attributed to users moving away from the page before snapshot/data being sent? Network issues? (interesting other hypotheses I didn't think about). etc. etc.

andyvan-ph commented:
@rcmarron I think we're closer than you think. This revision tightens things up quite nicely.

A couple of thoughts:

1) I think we could move the summary of improvements to screen recordings into the intro. In other words, we could start out by talking about the feature coming out beta, talking a little about how it works and how people use it.

2) Then we can go deeper into how we fixed this specific problem. This approach really fits into our aim of writing content for engineers, as [per our marketing personas](https://posthog.com/handbook/growth/marketing/customer-personas).

3) Neil's suggestion makes sense. I'd caution against going too in-depth, but would probably add a useful bit of context.

andyvan-ph commented on contents/blog/missing-recordings.md:
```diff
@@ -0,0 +1,73 @@
+---
+date: 2021-06-10
+title: The mystery of missing recordings at PostHog
```
```suggestion
title: How we fixed a major issue with our Session Recording beta
```

jamesefhawkins replied:
totally subjective (feel free to ignore) flyby: I think the former sounds like it'll be a more interesting read

andyvan-ph replied:
On reflection, you're totally the right

andyvan-ph commented on contents/blog/missing-recordings.md:
```diff
@@ -0,0 +1,73 @@
+One issue we faced was that we didn't want this feature to significantly increase the size of `posthog-js`, but we would have to include the placeholder image in the bundle. After some discussion, we landed on the idea of using a very small SVG that could easily scale to the size of the image being replaced. You can read more about the discussion that landed us here in [the PR itself](https://git ...
+
+
+After deploying the fix, the Sentry error went away, but the missing recording only went down a few percent.
```
```suggestion
After deploying the fix, two things happened:

1. The Sentry error went away 🥳
2. The missing recording only went down a few percent 😩

Back to the drawing board.
```

rcmarron replied:
Love the emojis!

andyvan-ph commented on contents/blog/missing-recordings.md:
```diff
@@ -0,0 +1,73 @@
+
+Why were the full snapshots missing? Presumably, if the later data was being sent up, the snapshot code must've already run. To further solidify this line of thinking, the code to determine the changes to the page depends entirely on the snapshot existing in memory, so the system must've already taken a full snapshot.
+
+After a day or so of pounding our heads on the keyboard, we realized that recordings were being split when a user was inactive for 30 minutes. But when the split happened, we were not taking a new full snapshot.
```
I feel like there's some missing detail here that would really round off the narrative. What steps / dead ends did we / you go down while you were pounding heads on keyboards?

andyvan-ph reviewed (commented):
Just gone through and made a few more suggestions. I think we're 95% there.

rcmarron commented:
Thanks for the feedback everyone! Super helpful! Just made another draft.

> @rcmarron did we use PostHog to measure the metrics? I wonder if the hook is going meta. Here's how we used PostHog to measure usage of PostHog and make Session Recordings awesome...

Sadly, it was mainly a metabase investigation.

> I'd love to see a quick summary of how session recording actually works with rrweb, what a full snapshot means, etc. etc.

Great idea! Added this in. It also gives us a cool opportunity to give credit to rrweb

> I'd also love hearing about predictions why the last 1% still exists: Can it all be attributed to users moving away from the page before snapshot/data being sent? Network issues? (interesting other hypotheses I didn't think about). etc. etc.

Love this too - added a quick sentence about it in.

> I think we could move the summary of improvements to screen recordings into the intro. In other words, we could start out by talking about the feature coming out beta, talking a little about how it works and how people use it.

Done! I think this helps the piece flow much nicer!

> Then we can go deeper into how we fixed this specific problem. This approach really fits into our aim of writing content for engineers, as [per our marketing personas](https://posthog.com/handbook/growth/marketing/customer-personas).

Agreed! Added some more detail here, and tried to make there be fewer logical jumps!

andyvan-ph commented on contents/blog/missing-recordings.md:
```diff
@@ -0,0 +1,98 @@
+---
+date: 2021-06-10
+title: The mystery of missing recordings at PostHog
```
```suggestion
title: The mystery of missing session recordings at PostHog
```

andyvan-ph commented:
@rcmarron has made some final commits to this and I think it's really great now. Great effort from him and thanks everyone who chipped in thoughts. Please give it a final look and suggest any final amends (I'm still thinking of possible tweaks to the title), but I may merge this tomorrow as I think it's ready.

andyvan-ph commented on contents/blog/missing-recordings.md:
```diff
@@ -0,0 +1,100 @@
+
+This is the story of how we decreased the number of missing recordings from 15% to <1%, and in doing so, moved the feature out of beta.
+
+>[Rick Marron](/handbook/company/team#rick-marron-software-engineer) is a software engineer at PostHog. In previous lives Rick was a Program Manager at Microsoft, and he spent over a year in Nairobi, Kenya where he developed the first publicly available real-time source for traffic data in the city using computer vision and street cameras. He's a native of Rochester, New York but lives in Los Ang ...
```
@rcmarron I've added this author bio (new thing I'm doing on the blog). Let me know if you're happy with it or change anything you don't like!

rcmarron replied:
Looks great to me!
